lib/membrane_video_compositor_multiple_inputs.ex

```diff
@@ -0,0 +1,294 @@
+
+    @spec has_frame?(__MODULE__.t()) :: boolean
+    def has_frame?(%__MODULE__{buffers: buffers}) do
+      not Enum.empty?(buffers)
+    end
+  end
+
+  @impl true
+  def handle_init(options) do
+    compositor_module = determine_compositor_module(options.implementation)
+
+    {:ok, internal_state} = compositor_module.init(options.caps)
+
+    state = %{
+      ids_to_tracks: %{},
+      caps: options.caps,
+      compositor_module: compositor_module,
+      internal_state: internal_state,
+      pads_to_ids: {0, %{}}
+    }
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_pad_added(pad, _context, state) do
+    state = register_track(state, pad)
+    {:ok, state}
+  end
+
+  defp register_track(state, pad) do
+    {new_id, pads_to_ids} = state.pads_to_ids
+    state = %{state | pads_to_ids: {new_id + 1, Map.put(pads_to_ids, pad, new_id)}}
+
+    %{state | ids_to_tracks: Map.put(state.ids_to_tracks, new_id, %Track{})}
+  end
+
+  @impl true
+  def handle_caps(pad, caps, _context, state) do
+    %{pads_to_ids: {_new_id, pads_to_ids}, internal_state: internal_state} = state
+    id = Map.get(pads_to_ids, pad)
+
+    {:ok, internal_state} =
+      state.compositor_module.add_video(id, caps, %{x: 0, y: 0}, internal_state)
+
+    state = %{state | internal_state: internal_state}
+    {{:ok, caps: {:output, state.caps}}, state}
+  end
+
+  @impl true
+  def handle_process(
+        pad,
+        buffer,
+        _context,
+        %{
+          ids_to_tracks: ids_to_tracks,
+          pads_to_ids: {_new_id, pads_to_ids}
+        } = state
+      ) do
+    id = Map.get(pads_to_ids, pad)
+
+    ids_to_tracks = push_frame(ids_to_tracks, id, buffer)
+
+    state = %{state | ids_to_tracks: ids_to_tracks}
+
+    case merge_frames(state) do
+      {{:merged, buffers}, state} -> {{:ok, buffer: {:output, buffers}}, state}
+      {{:empty, _empty}, state} -> {:ok, state}
+    end
+  end
+
+  defp merge_frames(
+         %{
+           ids_to_tracks: ids_to_tracks,
+           internal_state: internal_state
+         } = state
+       ) do
+    if all_have_frame?(ids_to_tracks) do
+      ids_to_frames = get_ids_to_frames(ids_to_tracks)
+
+      {{:ok, merged_frame_binary}, internal_state} =
+        state.compositor_module.merge_frames(ids_to_frames, internal_state)
+
+      ids_to_tracks = pop_frames(ids_to_tracks)
+
+      state = %{
+        state
+        | ids_to_tracks: ids_to_tracks,
+          internal_state: internal_state
+      }
+
+      state = remove_depleted_tracks(state)
+
+      {{_status, tail}, state} = merge_frames(state)
+      merged = [%Membrane.Buffer{payload: merged_frame_binary}] ++ tail
+
+      {{:merged, merged}, state}
+    else
+      {{:empty, []}, state}
+    end
+  end
+
+  defp push_frame(ids_to_tracks, id, frame) do
+    Map.update!(ids_to_tracks, id, fn track ->
+      Track.push_frame(track, frame)
+    end)
+  end
+
+  defp all_have_frame?(ids_to_tracks) when map_size(ids_to_tracks) == 0 do
+    false
+  end
```

Comment:
redundant

Reply:
Not exactly. From Elixir's documentation:
```
As the last example shows, Enum.all?/2 returns true if enumerable is empty,
regardless of fun. In an empty enumerable there is no element for which fun
returns a falsy value, so the result must be true. This is a well-defined
logical argument for empty collections.
```
tl;dr, `Enum.all?/2` will return `true` value for empty collection, but in our case, we want `false` result, if there are no tracks in the collection

Reply:
Hmm, so the name is kind of counterintuitive. Maybe do that check where the function is called

Reply:
Renamed